File: src/main/java/sc/fiji/gui/help/HelpManager.java
/*-
 * #%L
 * A framework for Java apps that attaches and shows simple help dialogs/wizzards to developer-chosen GUI controls.
 * %%
 * Copyright (C) 2024 Vladimir Ulman
 * %%
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 * #L%
 */
package sc.fiji.gui.help;

import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;

public class HelpManager {
	public static int HELP_KEY1 = KeyEvent.VK_H;
	public static int HELP_KEY2 = KeyEvent.VK_F1;

	public HelpManager(final Container theMainAppWindowContentPane) {
		this.theMainAppWindowContentPane = theMainAppWindowContentPane;

		//since keyboard event can arrive only to an element with focus,
		//we have to make sure the pane can be focused (can receive keyboard events)
		theMainAppWindowContentPane.setFocusable(true);
		theMainAppWindowContentPane.addKeyListener(new HelpKeyMonitor());

		//show no help when help key is pressed with mouse over the monitored main window,
		//it can be overridden later from the client code
		helpDialogs.put(theMainAppWindowContentPane, noHelpShower);

		//focuses this pane whenever mouse arrives over the app window (the main content pane),
		//which makes it see all keyboard events until the focus is changed (e.g. with mouse click or tab)
		doMonitorMouseOvers = true;
		theMainAppWindowContentPane.addMouseListener(new MouseListener() {
			@Override
			public void mouseEntered(MouseEvent e) {
				theMainAppWindowContentPane.requestFocusInWindow();
				itemWithMouseOver = theMainAppWindowContentPane;
			}
			@Override
			public void mouseClicked(MouseEvent e) {}
			@Override
			public void mousePressed(MouseEvent e) {}
			@Override
			public void mouseReleased(MouseEvent e) {}
			@Override
			public void mouseExited(MouseEvent e) {
				if (itemWithMouseOver == theMainAppWindowContentPane) {
					final int mouseX = e.getX();
					final int mouseY = e.getY();
					//is mouse leaving this (container) component through its outside boundary?
					//(in contrast to "leaving into" another (child) component that is inside/over
					// this (container) component)
					if (mouseX < 0 || mouseX >= theMainAppWindowContentPane.getWidth()
							|| mouseY < 0 || mouseY >= theMainAppWindowContentPane.getHeight()) {
						//left outside
						itemWithMouseOver = null;
					}
				}
			}
		});
	}

	public HelpManager() {
		doMonitorMouseOvers = false;
		theMainAppWindowContentPane = null;
	}

	private final Component theMainAppWindowContentPane;
	private Component itemWithMouseOver = null;
	private final MouseOverMonitor mouseOverListener = new MouseOverMonitor();
	private final boolean doMonitorMouseOvers;

	class MouseOverMonitor implements MouseListener {
		@Override
		public void mouseEntered(MouseEvent e) {
			itemWithMouseOver = e.getComponent();
		}
		@Override
		public void mouseExited(MouseEvent e) {
			if (itemWithMouseOver == e.getComponent()) itemWithMouseOver = null;
			//NB: to make sure only myself is removed
			//    (just in case somebody else managed to set this attrib already,
			//     which could happen when mouseEntered() on that somebody was called
			//     before mouseExited() of the actual mouse-over'ed item
			//     (theoretically, it shouldn't happen) )
		}
		@Override
		public void mouseClicked(MouseEvent e) {}
		@Override
		public void mousePressed(MouseEvent e) {}
		@Override
		public void mouseReleased(MouseEvent e) {}
	}

	class HelpKeyMonitor implements KeyListener {
		final Component monitoredComponent;
		public HelpKeyMonitor(final Component forThisComponent) {
			monitoredComponent = forThisComponent;
		}
		public HelpKeyMonitor() {
			monitoredComponent = null;
			//NB: flags that the currently (at the time this listener is triggered)
			//    mouse-over'ed component should be used instead of this.monitoredComponent
		}
		@Override
		public void keyTyped(KeyEvent e) {}
		@Override
		public void keyPressed(KeyEvent e) {}
		@Override
		public void keyReleased(KeyEvent e) {
			//TODO: this goes into the debug branch
			//System.out.println("key released: "+e.getKeyChar());
			if (e.getKeyCode() == HELP_KEY1 || e.getKeyCode() == HELP_KEY2) {
				if (monitoredComponent != null) showItemHelp(monitoredComponent);
				else if (itemWithMouseOver != null) showItemHelp(itemWithMouseOver);
			}
		}
	}

	// ==================================================================================================================
	public void registerComponentHelp(final Component guiComponent, final Path pathToLocalTopic, final String dialogTitle) {
		if (doMonitorMouseOvers) guiComponent.addMouseListener(mouseOverListener);
		guiComponent.addKeyListener(new HelpKeyMonitor(guiComponent));
		helpDialogs.put(guiComponent, new DefaultLocalHelpShower(pathToLocalTopic, dialogTitle));
	}

	public void registerComponentHelp(final Component guiComponent, final URL urlToRemoteTopic, final String dialogTitle) {
		if (doMonitorMouseOvers) guiComponent.addMouseListener(mouseOverListener);
		guiComponent.addKeyListener(new HelpKeyMonitor(guiComponent));
		helpDialogs.put(guiComponent, new DefaultRemoteHelpShower(urlToRemoteTopic, dialogTitle));
	}

	public void registerComponentHelp(final Component guiComponent, final HelpShower ownHelpDialog) {
		if (doMonitorMouseOvers) guiComponent.addMouseListener(mouseOverListener);
		guiComponent.addKeyListener(new HelpKeyMonitor(guiComponent));
		helpDialogs.put(guiComponent, ownHelpDialog);
	}

	public void registerComponentHelp(final Path pathToLocalTopic, final String dialogTitle) {
		if (theMainAppWindowContentPane != null)
			helpDialogs.put(theMainAppWindowContentPane, new DefaultLocalHelpShower(pathToLocalTopic, dialogTitle));
	}

	public void registerComponentHelp(final URL urlToRemoteTopic, final String dialogTitle) {
		if (theMainAppWindowContentPane != null)
			helpDialogs.put(theMainAppWindowContentPane, new DefaultRemoteHelpShower(urlToRemoteTopic, dialogTitle));
	}

	public void registerComponentHelp(final HelpShower ownHelpDialog) {
		if (theMainAppWindowContentPane != null) helpDialogs.put(theMainAppWindowContentPane, ownHelpDialog);
	}

	public static Path constructPathToLocalTopics(final Class<?> appClass, final String topic) {
		try {
			return Paths.get(appClass.getResource(topic+"/1.html").toURI()).getParent();
		} catch (URISyntaxException | NullPointerException e) {
			try {
				System.err.println("Failed finding the local help "+appClass.getSimpleName()+"/"+topic
						+", trying a default placeholder instead...");
				return Paths.get(HelpManager.class.getResource("defaultDescription.html").toURI());
				//NB: notice the name of this framework...
				//TODO: the defaultDescription.html could show a demo how to create such local help
			} catch (URISyntaxException | NullPointerException ex) {
				throw new RuntimeException("Requested help ("
						+appClass.getSimpleName()+"/"+topic+") as well as default substitute help was not found.");
			}
		}
	}

	/**
	 * An aider to construct URL objects without the hassle of dealing with the potential {@link MalformedURLException}.
	 * If invalid input is given, the methods return URL pointing at https://scijava.org/.
	 *
	 * @param urlAsPlainText URL string to be wrapped into a proper {@link URL} object.
	 * @return URL object wrapped around the textual URL.
	 */
	public static URL constructURL(final String urlAsPlainText) {
		try {
			return new URL(urlAsPlainText);
		} catch (MalformedURLException e) {
			try {
				return new URL("https://scijava.org/");
			} catch (MalformedURLException ex) {
				throw new RuntimeException("Total failure: Couldn't construct URL obj around simple valid URL string.");
			}
		}
	}

	// ==================================================================================================================
	/**
	 * Starts the help dialog for the given component if that component has been previously registered via
	 * the family of registering methods, such as {@link HelpManager#registerComponentHelp(Component, Path, String)}.
	 * If null is given, the method silently quits without showing anything.
	 *
	 * @param guiItem Component of which the help should be displayed.
	 */
	public void showItemHelp(Component guiItem) {
		if (guiItem == null) {
			//TODO: add log consumer to this class
			//TODO: this goes into the debug branch
			//System.err.println("would show help, but nothing is active!?");
			return;
		}
		//TODO: this goes into the debug branch
		//System.out.println("showing help for item "+guiItem);
		helpDialogs.get(guiItem).showNonModalHelpNow();
		//TODO: make resilient when the guiItem is by chance not found!
		//TODO: show it in a separate thread, non-modal
	}

	private final Map<Component, HelpShower> helpDialogs = new HashMap<>(10);
	private static final HelpShower noHelpShower = () -> { /* does nothing intentionally */ };
}

File: src/main/kotlin/sc/fiji/gui/help/HelpManager.kt
package sc.fiji.gui.help

import java.awt.Component
import java.awt.Desktop
import java.awt.MouseInfo
import java.awt.event.KeyEvent
import java.awt.event.KeyListener
import java.net.MalformedURLException
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*

/**
 * void HM.obtain().getKeyboardAction()
 * -- use for the binding+actions, triggers this manager
 * -- should be bound to "the usual place in the app", could be bound in multiple places
 * -- keyboard focus is (obviously) not altered
 *
 * void HM.obtain().registerComponentHelp(forThisComponent, ....help params....)
 * -- registers the component into the _sorted_ list of "help-enabled components"
 * -- when the manager is triggered, it attempts to find _the first_ mouse-over'ed component from the list
 */
object HelpManager {
    /**
     * Exposes the help manager's entry point (handle) to Action-oriented applications,
     * a method that's executed when the corresponding Action is triggered.
     * @return Reference on the right method from this help manager
     */
    val keyboardAction: Runnable
        get() = Runnable { processHelpKey() }

    /**
     * Creates [KeyListener], to be attached in the application, that
     * activates this help manager when any of the given keys are pressed.
     * @param watchForTheseKeys Set of keys that should trigger this help manager.
     * @return Always a new object that implements the [KeyListener].
     */
    fun getKeyboardListener(watchForTheseKeys: Set<Int>): KeyListener =
        HelpKeyListener(watchForTheseKeys)

    //
    private class HelpKeyListener(watchForTheseKeys: Set<Int>) : KeyListener {
        private val hotKeys: Set<Int> = HashSet(watchForTheseKeys)

        //
        override fun keyTyped(e: KeyEvent) {}
        override fun keyPressed(e: KeyEvent) {}
        @Synchronized
        override fun keyReleased(e: KeyEvent) {
            if (e.keyCode in hotKeys) processHelpKey()
        }
    }

    // ==================================================================================================================
    private infix fun Component.withHelp(helpDialog: HelpShower) = ComponentWithHelp(this, helpDialog)
    private class ComponentWithHelp(val component: Component, val helpDialog: HelpShower)

    private val helpDialogs: MutableList<ComponentWithHelp> = ArrayList(30)

    /**
     * Starts the help dialog for the given component if that component has been previously registered via
     * the family of registering methods, such as [HelpManager.registerComponentHelp].
     * If it wasn't registered or null is given, any particular dialog is thus not available, and nothing is
     * shown consequently leaving the call with false return value.
     *
     * @param guiItem Component of which the help should be displayed.
     * @return true when help dialog for the input component has been registered previously and was thus shown.
     */
    fun showHelpNow(guiItem: Component?): Boolean {
        if (guiItem == null) return false
        for (item in helpDialogs) {
            if (guiItem === item.component) {
                item.helpDialog.showNonModalHelpNow()
                return true
            }
        }
        return false
    }

    /**
     * Scans the registered GUI components to look for the first one under the mouse cursor.
     * The current state (visibility, position and size) of the components is considered,
     * as well as the current mouse position, naturally.
     *
     * This is the entry point that's called after triggering the local help in the client application.
     */
    private fun processHelpKey() {
        for (item in helpDialogs) {
            //Enable to debug:
            //System.out.println("Help: Considering component: " + item.component.getClass().getSimpleName()+", visible="+item.component.isShowing());
            if (item.component.isShowing && isCurrentMousePosOverComponent(item.component)) {
                //Enable to debug:
                //System.out.println("Would be now printing help for component: " + item.component.getClass().getSimpleName());
                item.helpDialog.showNonModalHelpNow()
                return
            }
        }
    }

    private fun isCurrentMousePosOverComponent(component: Component): Boolean {
        val p = MouseInfo.getPointerInfo().location
        val c = component.locationOnScreen
        p.translate(-c.x, -c.y) //NB: the same as ".sub(component.corner)"
        return !(p.x < 0 || p.y < 0 || p.x >= component.width || p.y >= component.height)
    }

    private fun Component.add(helpDialog: HelpShower) {
        val inArea = width * height
        var index = -1
        for (registeredHelp in helpDialogs) {
            val c = registeredHelp.component
            ++index

            //NB: considering the _current_ area of the (previously) registered components
            //    (not a stored area that might no longer be accurate),
            //    against the _currently_ added component
            val cArea = c.width * c.height
            if (inArea < cArea) {
                //'index' points now on the enlisted component that's larger
                helpDialogs.add(index, ComponentWithHelp(this, helpDialog))
                return
            }
        }
        //if we got here, there was no smaller registered component
        helpDialogs += ComponentWithHelp(this, helpDialog)
    }

    /**
     * Unregisters the given guiComponent together with its local help dialog.
     * It returns the status of the [::remove()][List] operation.
     * @param guiComponent The guiComponent that shall no longer provide a local help.
     * @return False if the given guiComponent was null or not present in the list, else True.
     */
    fun Component.unregisterHelp(): Boolean = helpDialogs.removeIf { it.component === this }

    // ==================================================================================================================
    infix fun Component.registerHelp(ownHelpDialog: HelpShower) = add(ownHelpDialog)

    @JvmOverloads
    fun registerComponentHelp(component: Component, pathToLocalTopic: Path, dialogTitle: String, startOnThisPageNumber: Int = 0) {
        //TODO: add the first start page number into the LocalHelpShower
        component.add(DefaultLocalHelpShower(pathToLocalTopic, dialogTitle))
    }

    fun registerComponentHelp(component: Component, urlToRemoteTopic: URL, dialogTitle: String) =
        component.add(DefaultRemoteHelpShower(urlToRemoteTopic, dialogTitle))

    fun registerComponentHelpForWebBrowser(component: Component, urlToRemoteTopic: URL) =
        component.add { openUrlInSystemBrowser(urlToRemoteTopic) }

    /**
     * An aider to obtain an absolute, local filesystem path to the resources folder of the provided class,
     * and this path is concatenated with a relative path to the 'topic'. Typically, a class is used that is responsible for
     * the functionality behind a GUI element for which [HelpManager.registerComponentHelp]
     * is called, and that carries the help resources with it. The path to the class resources folder is yielded by
     * querying for `appClass.getResource(topic+"/1.html")`.
     *
     * @param appClass The class whose resource folder is extracted.
     * @param topic The sub-folder in the resources folder is appended to the constructed path.
     * @return An absolute, local filesystem path to the topic.
     */
    inline infix fun <reified C> constructPathToLocalTopics(topic: String): Path {
        val appClass = C::class.java
        val error = "Requested help (${appClass.simpleName}/$topic) as well as default substitute help was not found."
        try {
            return Paths.get(appClass.getResource("$topic/1.html").toURI()).parent
        } catch (e: URISyntaxException) {
            try {
                System.err.println("Failed finding the local help " + appClass.simpleName + "/" + topic
                                   + ", trying a default placeholder instead...")
                return Paths.get(HelpManager::class.java.getResource("defaultDescription.html").toURI())
                //NB: notice the name of this framework...
            } catch (ex: URISyntaxException) {
                throw RuntimeException(error)
            } catch (ex: NullPointerException) {
                throw RuntimeException(error)
            }
        } catch (e: NullPointerException) {
            try {
                System.err.println("Failed finding the local help ${appClass.simpleName}/$topic, trying a default placeholder instead...")
                return Paths.get(HelpManager::class.java.getResource("defaultDescription.html").toURI())
            } catch (ex: URISyntaxException) {
                throw RuntimeException(error)
            } catch (ex: NullPointerException) {
                throw RuntimeException(error)
            }
        }
    }

    /**
     * An aider to construct URL objects without the hassle of dealing with the potential [MalformedURLException].
     * If invalid input is given, the methods return URL pointing at https://scijava.org/.
     *
     * @param urlAsPlainText URL string to be wrapped into a proper [URL] object.
     * @return URL object wrapped around the textual URL.
     */
    fun constructURL(urlAsPlainText: String): URL {
        return try {
            URI(urlAsPlainText).toURL()
        } catch (e: MalformedURLException) {
            try {
                URI("https://scijava.org/").toURL()
            } catch (ex: MalformedURLException) {
                throw RuntimeException("Total failure: Couldn't construct URL obj around simple valid URL string.")
            }
        }
    }


    /**
     * Opens OS-default web browser on the specified page.
     * @param url URL to the requested content.
     */
    fun openUrlInSystemBrowser(url: URL) = openUrlInSystemBrowser(url.toString())

    /**
     * Opens OS-default web browser on the specified page.
     * @param url URL to the requested content.
     */
    fun openUrlInSystemBrowser(url: String) {
        val myOS = System.getProperty("os.name").lowercase(Locale.getDefault())
        when {
            "mac" in myOS -> Runtime.getRuntime().exec(arrayOf("open", url))
            "nux" in myOS || "nix" in myOS -> Runtime.getRuntime().exec(arrayOf("xdg-open", url))
            Desktop.isDesktopSupported() -> Desktop.getDesktop().browse(URI(url))
            else -> println("Please, open this URL yourself: $url")
        }
    }
}